engine/verification/finder/engine.go

```diff
@@ -314,64 +341,199 @@ func (e *Engine) checkCachedReceipts() {
+			log.Debug().
+				Bool("removed", removed).
+				Msg("cached receipt has been removed")
 
-			// checks if the result has already been handled
+			// checks if the result has already been processed or discarded
 			if e.processedResultIDs.Has(resultID) {
 				log.Debug().Msg("drops handling already processed result")
 				return
 			}
-
-			// adds receipt to pending or ready mempools depending on its processable status
-			ready := e.isProcessable(&rdp.Receipt.ExecutionResult)
-			if ready {
-				// block for the receipt is available,
-				// receipt is ready for process
-				ok := e.readyReceipts.Add(rdp)
-				if !ok {
-					log.Debug().Msg("drops adding duplicate receipt to ready mempool")
-					return
-				}
-			} else {
-				// block for the receipt is not available
-				// receipt moves to pending mempool
-				ok := e.pendingReceipts.Add(rdp)
-				if !ok {
-					log.Debug().Msg("drops adding duplicate receipt to pending mempool")
-					return
-				}
-
-				// marks receipt pending for its block ID
-				err := e.pendingReceiptIDsByBlock.Append(rdp.Receipt.ExecutionResult.BlockID, receiptID)
-				if err != nil {
-					e.log.Error().
-						Err(err).
-						Hex("block_id", logging.ID(rdp.Receipt.ExecutionResult.BlockID)).
-						Hex("receipt_id", logging.ID(receiptID)).
-						Msg("could not append receipt to receipt-ids-by-block mempool")
-				}
+			if e.discardedResultIDs.Has(resultID) {
+				log.Debug().Msg("drops handling already discarded result")
+				return
 			}
 
-			// records the execution receipt id based on its result id
-			err := e.receiptIDsByResult.Append(resultID, receiptID)
-			if err != nil {
-				log.Debug().Err(err).Msg("could not add receipt id to receipt-ids-by-result mempool")
+			ready := e.isProcessable(&rdp.Receipt.ExecutionResult)
+			if !ready {
+				// adds receipt to pending mempool
+				e.addToPending(rdp)
 				return
 			}
 
-			log.Info().
-				Bool("ready", ready).
-				Msg("cached execution receipt moved to proper mempool")
+			// adds receipt to ready mempool
+			e.addToReady(rdp)
+		}()
+	}
+}
+
+// addToReady encapsulates the logic around adding a ReceiptDataPack to ready receipts mempool.
+// The ReceiptDataPack is however discarded it if finder engine is not staked at its block id.
+func (e *Engine) addToReady(receiptDataPack *verification.ReceiptDataPack) {
+	receiptID := receiptDataPack.Receipt.ID()
+	resultID := receiptDataPack.Receipt.ExecutionResult.ID()
+	blockID := receiptDataPack.Receipt.ExecutionResult.BlockID
+	log := e.log.With().
+		Hex("origin_id", logging.ID(receiptDataPack.OriginID)).
+		Hex("block_id", logging.ID(blockID)).
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).Logger()
+
+	// checks whether verification node is staked at snapshot of this result's block.
+	ok, err := e.stakedAtBlockID(blockID)
+	if err != nil {
+		log.Debug().Err(err).Msg("could not verify stake of verification node for result")
```

Comment:
We could simply by returning the error instead, and let the caller decide how to log it. Otherwise, we are [repeating the fields](https://github.com/onflow/flow-go/pull/245/files#diff-e2a1bac627623a345dc4e9caf84fa096cdecddbe535730904d7438062ffb3820R383-R387) to be logged.
```
func (e *Engine) addToReady(receiptDataPack *verification.ReceiptDataPack) error {
  ...
  if err != nil {
    return fmt.Errorf("could not verify stake of verification node for result: %w", err)
  }
}
```


same for `addToPending`, `pendingToReady`, and `discardPendingReceipts`

Reply:
done in https://github.com/onflow/flow-go/pull/245/commits/72ea9fe08525d46d18180a9c24ac15b26283f0b6 for `addToPending` and `addToReady`. The `pendingToReady` and `discardPendingReceipts`operate on a _set_ of receipts, hence it is not directly applicable to return an error, and returning a `multierror` may not be the best practice for that situation.